ggml-backend.c

```diff
@@ -0,0 +1,435 @@
+    } else {
+        // not ideal, but shouldn't be hit when copying from/to CPU
+        // TODO: print a performance warning in debug builds
+        size_t nbytes = ggml_nbytes(src);
+        void * data = malloc(nbytes);
+        ggml_backend_get_tensor(src, data, 0, nbytes);
+        ggml_backend_set_tensor(dst, data, 0, nbytes);
+        free(data);
+    }
+}
+
+// backend CPU
+
+struct ggml_backend_cpu_context {
+    int n_threads;
+    void * work_data;
+    size_t work_size;
+};
+
+static const char * ggml_backend_cpu_name(ggml_backend_context_t ctx) {
+    return "CPU";
+
+    UNUSED(ctx);
+}
+
+static void ggml_backend_cpu_free_context(ggml_backend_context_t ctx) {
+    struct ggml_backend_cpu_context * cpu_ctx = (struct ggml_backend_cpu_context *)ctx;
+    free(cpu_ctx->work_data);
+    free(ctx);
+}
+
+struct cpu_backend_buffer {
+    void * data;
+    size_t offset;
+    size_t size;
+};
+
+static const size_t TENSOR_ALIGNMENT = 64; // should be enough for AVX 512
+
+static size_t aligned_offset(const void * buffer, size_t offset, size_t alignment) {
+    assert(alignment && !(alignment & (alignment - 1))); // power of 2
+    size_t align = (alignment - (((uintptr_t)buffer + offset) % alignment)) % alignment;
+    return offset + align;
+}
+
+static ggml_backend_buffer_t ggml_backend_cpu_alloc_buffer(ggml_backend_context_t ctx, size_t size) {
+    struct cpu_backend_buffer * buffer = malloc(sizeof(struct cpu_backend_buffer));
+    buffer->data = malloc(size);
+    buffer->offset = aligned_offset(buffer->data, 0, TENSOR_ALIGNMENT);
+    buffer->size = size;
+    return buffer;
+
+    UNUSED(ctx);
+}
+
+static void ggml_backend_cpu_free_buffer(ggml_backend_context_t ctx, ggml_backend_buffer_t buffer) {
+    struct cpu_backend_buffer * cpu_buffer = (struct cpu_backend_buffer *)buffer;
+    free(cpu_buffer->data);
+    free(cpu_buffer);
+
+    UNUSED(ctx);
+}
+
+static void ggml_backend_cpu_reset_buffer(ggml_backend_context_t ctx, ggml_backend_buffer_t buffer) {
+    struct cpu_backend_buffer * cpu_buffer = (struct cpu_backend_buffer *)buffer;
+    cpu_buffer->offset = aligned_offset(cpu_buffer->data, 0, TENSOR_ALIGNMENT);
+
+    UNUSED(ctx);
+}
+
+static void ggml_backend_cpu_alloc_tensor(ggml_backend_context_t ctx, ggml_backend_buffer_t buffer, struct ggml_tensor * tensor) {
+    struct cpu_backend_buffer * cpu_buffer = (struct cpu_backend_buffer *)buffer;
+
+    // TODO: make this error recoverable
+    if (cpu_buffer->offset + ggml_nbytes(tensor) > cpu_buffer->size) {
+        fprintf(stderr, "%s: not enough space in the buffer (needed %zu, available %zu)\n",
+                __func__, ggml_nbytes(tensor), cpu_buffer->size - cpu_buffer->offset);
+        GGML_ASSERT(false);
+    }
+
+    tensor->data = (char*)cpu_buffer->data + cpu_buffer->offset;
+    cpu_buffer->offset = aligned_offset(cpu_buffer->data, cpu_buffer->offset + ggml_nbytes(tensor), TENSOR_ALIGNMENT);
```

Comment:
That looks interesting, but I imagine that you would have to use it in the kernels for the compiler to actually notice anything. It may be worth checking if adding that to all the data pointers in the kernels improves performance. The tensor allocator in the ggml-cuda backend always aligns pointers 128 bytes, so it is safe to assume that tensor data pointers are aligned to at least that.

Reply:
About using this on the CPU code, there are AVX instructions both for aligned and unaligned load and stores. We always use the unaligned instructions because in practice in current CPUs it doesn't seem to make much of a different in performance, and depending on the types and row sizes the data may not always be aligned.